sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala

```diff
@@ -389,6 +389,25 @@ class TypeCoercionSuite extends AnalysisTest {
     widenTest(StringType, MapType(IntegerType, StringType, true), None)
     widenTest(ArrayType(IntegerType), StructType(Seq()), None)
 
+    widenTest(
+      ArrayType(StringType, containsNull=true),
+      ArrayType(StringType, containsNull=false),
+      Some(ArrayType(StringType, containsNull=true)))
+    widenTest(
+      MapType(StringType, StringType, valueContainsNull=true),
+      MapType(StringType, StringType, valueContainsNull=false),
+      Some(MapType(StringType, StringType, valueContainsNull=true)))
```

Comment:
please add a test also for struct

Reply:
Here are tests for nested complex types, https://github.com/apache/spark/pull/20010/files#diff-01ecdd038c5c2f53f38118912210fef8R405